File: benchmark/nodejs/spectral_norm.js
function eval_A(i, j) {
    return 1.0 / (((i + j) * (i + j + 1) / 2) + i + 1);
}

function eval_A_times_u(u, v, n) {
    for (let i = 0; i < n; i++) {
        v[i] = 0.0;
        for (let j = 0; j < n; j++) {
            v[i] += eval_A(i, j) * u[j];
        }
    }
}

function eval_At_times_u(u, v, n) {
    for (let i = 0; i < n; i++) {
        v[i] = 0.0;
        for (let j = 0; j < n; j++) {
            v[i] += eval_A(j, i) * u[j];
        }
    }
}

function eval_AtA_times_u(u, v, w, n) {
    eval_A_times_u(u, w, n);
    eval_At_times_u(w, v, n);
}

const start = Date.now();
const n = 150;
let u = new Float64Array(n);
let v = new Float64Array(n);
let w = new Float64Array(n);

for (let i = 0; i < n; i++) u[i] = 1.0;

for (let i = 0; i < 10; i++) {
    eval_AtA_times_u(u, v, w, n);
    eval_AtA_times_u(v, u, w, n);
}

let vBv = 0.0;
let vv = 0.0;
for (let i = 0; i < n; i++) {
    vBv += u[i] * v[i];
    vv += v[i] * v[i];
}

const res = Math.sqrt(vBv / vv);
const end = Date.now();

console.log(`Result: ${res.toFixed(9)}`);
console.log(`Time: ${end - start}ms`);

File: benchmarks/src/lua/spectral_norm.lua
local function eval_A(i, j)
    return 1.0 / (((i + j) * (i + j + 1.0) / 2.0) + i + 1.0)
end

local function eval_A_times_u(u, v, n)
    for i = 0, n - 1 do
        v[i] = 0.0
        for j = 0, n - 1 do
            v[i] = v[i] + eval_A(i, j) * u[j]
        end
    end
end

local function eval_At_times_u(u, v, n)
    for i = 0, n - 1 do
        v[i] = 0.0
        for j = 0, n - 1 do
            v[i] = v[i] + eval_A(j, i) * u[j]
        end
    end
end

local function eval_AtA_times_u(u, v, w, n)
    eval_A_times_u(u, w, n)
    eval_At_times_u(w, v, n)
end

local start = os.clock() * 1000
local n = 150
local u, v, w = {}, {}, {}
for i = 0, n - 1 do
    u[i] = 1.0
    v[i] = 0.0
    w[i] = 0.0
end

for i = 1, 10 do
    eval_AtA_times_u(u, v, w, n)
    eval_AtA_times_u(v, u, w, n)
end

local vBv = 0.0
local vv = 0.0
for i = 0, n - 1 do
    vBv = vBv + u[i] * v[i]
    vv = vv + v[i] * v[i]
end

local res = math.sqrt(vBv / vv)
local end_time = os.clock() * 1000

print(string.format("Result: %.9f", res))
print("Time: " .. math.floor(end_time - start) .. "ms")
